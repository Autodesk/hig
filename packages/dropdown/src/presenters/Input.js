import React from "react";
import Icon from "@hig/icon";
import { TextFieldPresenter } from "@hig/text-field";
import "@hig/icon/build/index.css";
import "@hig/text-field/build/index.css";

import "./Input.scss";

export default function Input(props) {
  return (
    <div className="hig__dropdown__input-wrapper">
      <TextFieldPresenter {...props} type="text" readOnly />
      <span className="hig__dropdown__input-caret">
        {/* @TODO: there are variations of the TextField with multiple icons at the end of the input. These icon nodes should be passed as props to TextField. */}
        <Icon name="caret" />
      </span>
    </div>
  );
}

Input.propTypes = {
  ...TextFieldPresenter.propTypes,
  type: undefined,
  readOnly: undefined
};
