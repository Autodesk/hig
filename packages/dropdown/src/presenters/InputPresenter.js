import React, { Component } from "react";
import Icon from "@hig/icon";
import { TextFieldPresenter } from "@hig/text-field";
import "@hig/icon/build/index.css";
import "@hig/text-field/build/index.css";
import PropTypes from "prop-types";

import "./InputPresenter.scss";

export default class InputPresenter extends Component {
  static propTypes = {
    /**
     * Prevents user actions on the field
     */
    disabled: PropTypes.bool,
    /**
     * Error text for the field. Setting this value applies error styling to the entire component.
     */
    errors: PropTypes.string,
    /**
     * Allows managing field's focus via prop. Setting this value overrides the internal focus state management.
     */
    focused: PropTypes.bool,
    /**
     * Prevents user actions on the field
     */
    hasTextEntry: PropTypes.bool,
    /**
     * When true, displays passed error text. When false, displays instructions with error styling.
     */
    hideInstructionsOnErrors: PropTypes.bool,
    /**
     * HTML ID attribute
     */
    id: PropTypes.string,
    /**
     * Icon element that precedes the input.
     */
    icon: PropTypes.node,
    /**
     * Instructional text for the field
     */
    instructions: PropTypes.string,
    /**
     * Text describing what the field represents
     */
    label: PropTypes.string,
    /**
     * Name of the field when submitted with a form
     */
    name: PropTypes.string,
    /**
     * Called when user moves focus from the field
     */
    onBlur: PropTypes.func,
    /**
     * Called after user changes the value of the field
     */
    onChange: PropTypes.func,
    /**
     * Called when user clicks the clear button
     */
    onClearButtonClick: PropTypes.func,
    /**
     * Called when user puts focus onto the field
     */
    onFocus: PropTypes.func,
    /**
     * Called as user changes the value of the field
     */
    onInput: PropTypes.func,
    /**
     * Example of what the user should type into the field
     */
    placeholder: PropTypes.string,
    /**
     * Text describing why the field is required
     */
    required: PropTypes.string,
    /**
     * When true, causes the clear button to appear
     */
    showClearButton: PropTypes.bool,
    /**
     * Value of the field
     */
    value: PropTypes.string
  };

  render() {
    let textFieldPresenter;
    const { hasTextEntry, ...inputProps } = this.props;

    if (hasTextEntry) {
      textFieldPresenter = <TextFieldPresenter {...inputProps} type="text" />;
    } else {
      textFieldPresenter = (
        <TextFieldPresenter {...inputProps} type="button" readOnly />
      );
    }

    return (
      <div className="hig__dropdown__input-wrapper">
        {textFieldPresenter}
        <span className="hig__dropdown__input-caret">
          {/* @TODO: there are variations of the TextField with multiple icons at the end of the input. These icon nodes should be passed as props to TextField. */}
          <Icon name="caret" />
        </span>
      </div>
    );
  }
}
