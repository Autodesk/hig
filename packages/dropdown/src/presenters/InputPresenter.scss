@import "../tokens.scss";

.hig__dropdown__input-wrapper {
  position: relative;

  // TODO: Refactor TextField to avoid needing to override styles
  .hig__text-field__input {
    font-weight: 600;
  }

  // TODO: Refactor TextField to avoid needing to override styles
  .hig__text-field__extra--dropdown-caret {
    bottom: 6px;
  }
}

.hig__dropdown__input-caret {
  position: absolute;
  // 30px = the height of the icon element
  top: calc(#{$menu-top-offset} - 30px);
  right: 7px;
  z-index: -1;

  .hig__dropdown--disabled & {
    opacity: 0.3;
  }
}
