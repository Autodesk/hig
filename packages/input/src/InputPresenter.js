import React from "react";
import PropTypes from "prop-types";
import { css } from "emotion";

import { ThemeContext } from "@hig/themes";

import stylesheet from "./stylesheet";

export default function InputPresenter({
  disabled,
  hasFocus,
  hasHover,
  value,
  onFocus,
  onBlur,
  onMouseEnter,
  onMouseLeave,
  onChange
}) {
  return (
    <ThemeContext.Consumer>
      {({ themeData }) => {
        const styles = stylesheet(
          { isDisabled: disabled, hasFocus, hasHover },
          themeData
        );

        return (
          <div className={css(styles.inputWrapper)}>
            <input
              className={css(styles.input)}
              value={value}
              disabled={disabled}
              onFocus={onFocus}
              onBlur={onBlur}
              onMouseEnter={onMouseEnter}
              onMouseLeave={onMouseLeave}
              onChange={onChange}
            />
            <div className={css(styles.halo)} />
          </div>
        );
      }}
    </ThemeContext.Consumer>
  );
}

InputPresenter.propTypes = {
  disabled: PropTypes.bool,
  hasFocus: PropTypes.bool,
  hasHover: PropTypes.bool,
  value: PropTypes.string,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
  onChange: PropTypes.func
};
