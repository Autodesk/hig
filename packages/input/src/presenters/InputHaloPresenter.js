import React from "react";
import PropTypes from "prop-types";
import { cx, css } from "emotion";

import { ThemeContext } from "@hig/theme-context";

import stylesheet from "./InputHaloPresenter.stylesheet";
import { availableVariants } from "../constants";

function InputHaloPresenter(props) {
  const {
    children,
    hasFocus,
    hasHover,
    isDisabled,
    stylesheet: customStylesheet,
    variant,
    error,
    ...otherProps
  } = props;

  return (
    <ThemeContext.Consumer>
      {({ resolvedRoles, metadata }) => {
        const styles = stylesheet(
          { isDisabled, hasFocus, hasHover, variant, error },
          resolvedRoles
        );
        const cssStyles = customStylesheet
          ? customStylesheet(
              styles,
              props,
              resolvedRoles,
              metadata.colorSchemeId
            )
          : styles;

        const { className } = otherProps;

        return (
          <div className={cx(css(cssStyles.wrapper), className)}>
            {children}
            <div
              className={cx(
                css(cssStyles.halo),
                className
                  ? className
                      .split(" ")
                      .reduce((acc, cur) => cx(acc, `${cur.trim()}__halo`), "")
                  : ""
              )}
            />
          </div>
        );
      }}
    </ThemeContext.Consumer>
  );
}

InputHaloPresenter.propTypes = {
  children: PropTypes.node,
  hasFocus: PropTypes.bool,
  hasHover: PropTypes.bool,
  isDisabled: PropTypes.bool,
  stylesheet: PropTypes.func,
  variant: PropTypes.oneOf(availableVariants),
  error: PropTypes.bool
};

export default InputHaloPresenter;
