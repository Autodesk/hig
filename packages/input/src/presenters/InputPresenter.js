import React from "react";
import PropTypes from "prop-types";
import { cx, css } from "emotion";

import { ThemeContext } from "@hig/theme-context";

import stylesheet from "./InputPresenter.stylesheet";
import {
  availableTagNames,
  availableVariants,
  tagNames,
  variants
} from "../constants";

export default function InputPresenter(props) {
  const {
    disabled,
    hasFocus,
    hasHover,
    inputRef,
    onBlur,
    onFocus,
    onMouseEnter,
    onMouseLeave,
    tagName,
    stylesheet: customStylesheet,
    variant,
    ...otherProps
  } = props;
  return (
    <ThemeContext.Consumer>
      {({ resolvedRoles, metadata }) => {
        const styles = stylesheet(
          { isDisabled: disabled, hasFocus, hasHover, variant },
          resolvedRoles
        );
        const cssStyles = customStylesheet
          ? customStylesheet(
              styles,
              props,
              resolvedRoles,
              metadata.colorSchemeId
            )
          : styles;

        let className = css(cssStyles.input);
        const { className: customClassName } = otherProps;
        if (customClassName) {
          className = cx(
            className,
            variant === variants.PLAIN
              ? customClassName
              : customClassName
                  .split(" ")
                  .reduce((acc, cur) => cx(acc, `${cur.trim()}__input`), "")
          );
        }

        const tagNameKey = tagName.toUpperCase();
        const Element = tagNames[tagNameKey];

        return (
          <Element
            {...otherProps}
            className={className}
            disabled={disabled}
            onBlur={onBlur}
            onFocus={onFocus}
            onMouseEnter={onMouseEnter}
            onMouseLeave={onMouseLeave}
            ref={inputRef}
          />
        );
      }}
    </ThemeContext.Consumer>
  );
}

InputPresenter.propTypes = {
  disabled: PropTypes.bool,
  hasFocus: PropTypes.bool,
  hasHover: PropTypes.bool,
  onBlur: PropTypes.func,
  onFocus: PropTypes.func,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
  stylesheet: PropTypes.func,
  tagName: PropTypes.oneOf(availableTagNames),
  variant: PropTypes.oneOf(availableVariants)
};

InputPresenter.defaultProps = {
  tagName: tagNames.INPUT
};
