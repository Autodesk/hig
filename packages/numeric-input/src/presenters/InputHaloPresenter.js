import React from "react";
import PropTypes from "prop-types";
import { css } from "emotion";

import { ThemeContext } from "@hig/theme-context";

import stylesheet from "./InputHaloPresenter.stylesheet";
import { availableVariants } from "../constants";

function InputHaloPresenter(props) {
  const {
    children,
    hasFocus,
    hasHover,
    isDisabled,
    stylesheet: customStylesheet,
    variant
  } = props;

  return (
    <ThemeContext.Consumer>
      {({ resolvedRoles, metadata }) => {
        const styles = stylesheet(
          { isDisabled, hasFocus, hasHover, variant },
          resolvedRoles
        );
        const cssStyles = customStylesheet
          ? customStylesheet(
              styles,
              props,
              resolvedRoles,
              metadata.colorSchemeId
            )
          : styles;

        return (
          <div className={css(cssStyles.wrapper)}>
            {children}
            <div className={css(cssStyles.halo)} />
          </div>
        );
      }}
    </ThemeContext.Consumer>
  );
}

InputHaloPresenter.propTypes = {
  children: PropTypes.node,
  hasFocus: PropTypes.bool,
  hasHover: PropTypes.bool,
  isDisabled: PropTypes.bool,
  stylesheet: PropTypes.func,
  variant: PropTypes.oneOf(availableVariants)
};

export default InputHaloPresenter;
