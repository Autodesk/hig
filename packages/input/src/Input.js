import React from "react";
import PropTypes from "prop-types";
import { ControlBehavior } from "@hig/behaviors";

import InputPresenter from "./presenters/InputPresenter";
import InputHaloPresenter from "./presenters/InputHaloPresenter";

import {
  availableTagNames,
  availableVariants,
  tagNames,
  variants
} from "./constants";

function Wrapper(props) {
  if (props.variant === variants.PLAIN) {
    return props.children;
  }

  return <InputHaloPresenter {...props} />;
}

Wrapper.propTypes = {
  children: PropTypes.node,
  tagName: PropTypes.oneOf(availableTagNames),
  variant: PropTypes.oneOf(availableVariants)
};

function Input(props) {
  const {
    onFocus: onFocusProp,
    onBlur: onBlurProp,
    onMouseEnter: onMouseEnterProp,
    onMouseLeave: onMouseLeaveProp,
    disabled: disabledProp,
    inputRef,
    stylesheet,
    tagName,
    variant,
    error,
    ...otherProps
  } = props;

  return (
    <ControlBehavior
      onFocus={onFocusProp}
      onBlur={onBlurProp}
      onMouseEnter={onMouseEnterProp}
      onMouseLeave={onMouseLeaveProp}
    >
      {({
        hasHover,
        onMouseEnter,
        onMouseLeave,
        hasFocus,
        onFocus,
        onBlur
      }) => (
        <Wrapper
          isDisabled={disabledProp}
          hasFocus={hasFocus}
          hasHover={hasHover}
          stylesheet={stylesheet}
          tagName={tagName}
          variant={variant}
          error={error}
          {...otherProps}
        >
          <InputPresenter
            disabled={disabledProp}
            hasFocus={hasFocus}
            hasHover={hasHover}
            inputRef={inputRef}
            onBlur={onBlur}
            onFocus={onFocus}
            onMouseEnter={onMouseEnter}
            onMouseLeave={onMouseLeave}
            stylesheet={stylesheet}
            tagName={tagName}
            variant={variant}
            {...otherProps}
          />
        </Wrapper>
      )}
    </ControlBehavior>
  );
}

Input.propTypes = {
  /**
   * The initial value of the control
   */
  defaultValue: PropTypes.string,
  /**
   * Prevents the user from interacting with the input
   */
  disabled: PropTypes.bool,
  /**
   * A callback ref that gets passed to the HTML input
   */
  inputRef: PropTypes.func,
  /**
   * Fired when an element has lost focus
   */
  onBlur: PropTypes.func,
  /**
   * Fired when an element has received focus
   */
  onFocus: PropTypes.func,
  /**
   * Fired when a pointing device is moved over the element
   */
  onMouseEnter: PropTypes.func,
  /**
   * Fired when the pointer of a pointing device is moved out of an element
   */
  onMouseLeave: PropTypes.func,
  /**
   * Adds custom/overriding styles
   */
  stylesheet: PropTypes.func,
  /**
   * Renders the HTML element that is passed in
   */
  tagName: PropTypes.oneOf(availableTagNames),
  /**
   * The value of the control
   */
  value: PropTypes.string,
  /**
   * The visual variant of the input
   */
  variant: PropTypes.oneOf(availableVariants),
  /**
   * Specify if the value of the input is wrong
   */
  error: PropTypes.bool
};

Input.defaultProps = {
  tagName: tagNames.INPUT,
  variant: variants.LINE,
  error: false
};

export default Input;
