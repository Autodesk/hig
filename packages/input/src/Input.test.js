import React from "react";
import { mount } from "enzyme";
import {
  behavesLikeFocusBehavior,
  behavesLikeHoverBehavior,
  composesLikeControlBehavior
} from "@hig/behaviors/test";

import Input from "./Input";
import InputPresenter from "./presenters/InputPresenter";
import InputHaloPresenter from "./presenters/InputHaloPresenter";

describe("Input", () => {
  behavesLikeFocusBehavior(<Input />);
  behavesLikeHoverBehavior(<Input />);
  composesLikeControlBehavior({
    Subject: Input,
    Receiver: InputPresenter
  });

  describe("rendering variants", () => {
    describe("when variant='line'", () => {
      it("renders the line-style input", () => {
        const wrapper = mount(<Input variant="line" />);

        const decoration = wrapper.find(InputHaloPresenter);
        expect(decoration).toHaveProp("variant", "line");
      });
    });

    describe("when variant='box'", () => {
      it("renders the box-style  input", () => {
        const wrapper = mount(<Input variant="box" />);

        const decoration = wrapper.find(InputHaloPresenter);
        expect(decoration).toHaveProp("variant", "box");
      });
    });

    describe("when variant='plain'", () => {
      it("renders a plain input", () => {
        const wrapper = mount(<Input variant="plain" />);

        expect(wrapper).not.toContain(InputHaloPresenter);
      });
    });
  });

  describe("when controlled", () => {
    let onChangeSpy;
    let wrapper;
    let interactiveElement;

    beforeEach(() => {
      onChangeSpy = jest.fn();
      wrapper = mount(<Input value="foo" onChange={onChangeSpy} />);
      interactiveElement = wrapper.find("input");
    });

    it("sets the value", () => {
      expect(interactiveElement).toHaveProp("value", "foo");
    });

    describe("typing into the field", () => {
      beforeEach(() => {
        interactiveElement.getDOMNode().value = "bar";
        interactiveElement.simulate("change");
      });

      it("calls the callback", () => {
        expect(onChangeSpy.mock.calls.length).toEqual(1);
      });

      describe("the interactive element", () => {
        it("keeps the value from the value prop", () => {
          expect(interactiveElement.getDOMNode().value).toEqual("foo");
        });
      });
    });
  });

  describe("when uncontrolled", () => {
    let onChangeSpy;
    let wrapper;
    let interactiveElement;

    beforeEach(() => {
      onChangeSpy = jest.fn();
      wrapper = mount(<Input defaultValue="foo" onChange={onChangeSpy} />);
      interactiveElement = wrapper.find("input");
    });

    it("sets the defaultValue", () => {
      expect(interactiveElement).toHaveProp("defaultValue", "foo");
    });

    describe("typing into the field", () => {
      beforeEach(() => {
        interactiveElement.getDOMNode().value = "bar";
        interactiveElement.simulate("change");
      });

      it("calls the callback", () => {
        expect(onChangeSpy.mock.calls.length).toEqual(1);
      });

      describe("the interactive element", () => {
        it("gets the typed value", () => {
          expect(interactiveElement.getDOMNode().value).toEqual("bar");
        });
      });
    });
  });

  it("passes arbitrary props to input element", () => {
    const wrapper = mount(<Input data-my-attr="foo" />);
    expect(wrapper.find("input")).toHaveProp("data-my-attr", "foo");
  });

  it("passes className prop to input element", () => {
    const className = "custom-class";
    const wrapper = mount(<Input className={className} />);
    expect(wrapper.hasClass(className)).toBe(true);
    expect(wrapper.find("input").hasClass(`${className}__input`)).toBe(true);
  });
});
