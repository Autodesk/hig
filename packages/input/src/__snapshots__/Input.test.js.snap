// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="css-a2gl53"
>
  <input
    className="css-93ae2i"
    disabled={true}
    onBlur={[Function]}
    onChange={undefined}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    value="bran.stark@winterfell.com"
  />
  <div
    className="css-1e2ir1l"
  />
</div>
`;

exports[`Input renders children 1`] = `
<div
  className="css-a2gl53"
>
  <input
    className="css-kqzj2d"
    disabled={undefined}
    onBlur={[Function]}
    onChange={undefined}
    onFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    value="jon.snow@winterfell.com"
  />
  <div
    className="css-v5j8f1"
  />
</div>
`;
